nipc-openapi/NIPC.yaml

```diff
@@ -0,0 +1,1619 @@
+        property registration. You cannot read from a group id.
+      operationId: ActionEventStart
+      parameters:
+        - name: id
+          in: path
+          description: device id that need to be filtered, group id not allowed
+          required: true
+          schema:
+            type: string
+            format: uuid
+            example: 12345678-1234-5678-1234-56789abcdef4
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: './protocolmaps/ProtocolMap.yaml#/components/schemas/ProtocolMap-Property'
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SuccessResponse'          
+        '400':
+          description: Bad request
+        '401':
+          description: Unauthorized
+        '405':
+          description: Invalid request 
+        '500':
+          description: Server-side failure
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/FailureResponse' 
+                  - $ref: '#/components/schemas/EventErrorCodes'
+  
+  /{id}/action/event/stop:
+    post:
+      tags:
+        - NIPC action APIs with embedded protocol mapping
+      summary: Tell a device to stop reporting an event using a protocol mapping
+      description: |-
+        Tell a device to stop reporting an event, embedding property protocol mapping in the API, this does not require
+        property registration. You cannot read from a group id.
+      operationId: ActionEventStop
+      parameters:
+        - name: id
+          in: path
+          description: device id that need to be filtered, group id not allowed
+          required: true
+          schema:
+            type: string
+            format: uuid
+            example: 12345678-1234-5678-1234-56789abcdef4
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: './protocolmaps/ProtocolMap.yaml#/components/schemas/ProtocolMap-Property'
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SuccessResponse'          
+        '400':
+          description: Bad request
+        '401':
+          description: Unauthorized
+        '405':
+          description: Invalid request 
+        '500':
+          description: Server-side failure
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/FailureResponse' 
+                  - $ref: '#/components/schemas/EventErrorCodes'
+ 
+### Registrations
+  /registration/sdfobject/{sdfobject}:
+    post:
```

Comment:
Is POST method needed if the resulting resource URL is already known or would PUT be sufficient?

Reply:
I would expect the POST and PUT to behave identically here but I am not entirely sure what the `{sdfobject}` will be in the path. 
With the example sdf object, will it be: 
`/registration/sdfobject/https%3A%2F%2Fexample.com%2Fheartrate%23%2FsdfObject%2Fhealthsensor`

If this is the case, we could have a `POST /registration/sdfobject` instead since that uri can always be calculated from the body.
@bbrinckman thoughts?

Reply:
Since we are creating the object in the NIPC Gateway, I thought it would be a post. A put would look exactly the same. @rohitmohan96 you are right {sdfobject} is not really required, as it is already referenced in the body

Reply:
Then perhaps POST to `/registration/sdfobject/` is good way to store new models and PUT to the model specific URL (also returned by the POST) the way to update.

Reply:
And perhaps "models" would be more appropriate URL path component name (instead of "sdfobject") for the collection resource. And since the payload would be standard SDF, you can change the `content` from `application/json` into `application/sdf+json`

Reply:
So it would be `POST /{id}/registration/model` (where id is the device id)?

Reply:
I updated this API to move {id} to the path.
POST creates the model and returns it togethter with the id it was registered for. We return the entire model, which implies the URL. We could also just return the URL. @rohitmohan96 @akeranen any thoughts on the response?
PUT does take {sdfobject} as a parameter, to indicate with object to modify. GET/DELETE also take this parameter

Reply:
> So it would be `POST /{id}/registration/model` (where id is the device id)?

Something like this, yes (maybe consider some path components before the ID). And I think it makes sense to have also a non-ID specific version of the API endpoint to register models without binding to a device ID so that you can pre-populate the GW before the devices show up. 

Then (after using either API to register new model) you could refer to the models in ID-specific registrations, like

```
POST /{id}/registration/definition
Payload: https://example.com/heartrate#sdfObject/healthsensor
```

Reply:
> I updated this API to move {id} to the path. POST creates the model and returns it togethter with the id it was registered for. We return the entire model, which implies the URL. We could also just return the URL. @rohitmohan96 @akeranen any thoughts on the response? PUT does take {sdfobject} as a parameter, to indicate with object to modify. GET/DELETE also take this parameter

Returning the URL of the new resource in a Location header is good practice. Returning the model as already submitted maybe isn't that useful. But since there are now two URLs(?), the device-ID-specific and generic (without device ID), maybe returning one in the Location header and another in payload could make sense (haven't thought about this in detail though).

Reply:
I still do not understand why we would register a model without a device ID. I see we can re-use a model, but would we register a model that does not apply to a device?

Reply:
You could register a set of models **before** registering the device. Those could be, for example, different tools or users who have the responsibility to do that.

Reply:
Yes, I think the most convenient is to include the model or a reference to the model in the SCIM request. 
NIPC does need to allow registration of an SDF model by the control app though, as these would be different entities.

WRT registering the model without an id, would it not be convenient then to leverage a group id, even without any members? You can then add members later.

Reply:
Do you have an example of how would this look like in a SCIM request? Keeping the model registration for the NIPC API seems like a good idea. The group id use case seems reasonable but is there a problem registering a model without any ID that I'm missing here?